A few more changes to the message bars component

julienw opened the pull request:
This is a follow-up to my previous patch for the message bars. When I started using them I realized they had a few problems that I try to fix it, hopefully completely this time.

* the icon now stays at the top when the text is multiline.
* better behavior when the available space is small: buttons don't shrink anymore, and the closing button always has some margin.
* added the "generic" style.

In a future patch (maybe a "good first issue" :) ), I'd like to use the generic style for the service worker notice. In the past I did this notice following photon guidelines but without thinking of making it reusable.

It might be easier to look at the individual commits.

julienw commented on package.json:
```diff
@@ -13,7 +13,7 @@
     "lint-js": "eslint *.js bin src",
     "lint-fix-js": "yarn lint-js --fix",
-    "lint-css": "stylelint 'src/**/*.css' 'res/css/**/*.css'",
+    "lint-css": "stylelint 'src/**/*.css' 'res/**/*.css'",
```
I realized that the css for the photon example page wasn't checked. I did it in the first commit that you might want to look separately.

gregtatum replied:
Ah, whoops! Thanks for the fix here.

julienw commented on res/css/photon/message-bar.css:
```diff
@@ -5,53 +5,66 @@
 /* See https://design.firefox.com/photon/components/message-bars.html */
 .photon-message-bar {
   display: flex;
-  width: 100%;
```
I removed this because this was in the way of using it in a flex container to make it adjust to its content.
In existing usages it was a noop because divs always take the full width.

julienw commented on res/css/photon/message-bar.css:
```diff
@@ -5,53 +5,66 @@
   min-height: 32px;
   box-sizing: border-box;
   align-items: center;
-  padding: 0 4px;
```
This padding was useful when we weren't having an icon at the left, but now we always have one.

julienw commented on res/css/photon/message-bar.css:
```diff
@@ -5,53 +5,66 @@
-.photon-message-bar .photon-button {
-  margin-left: 8px;
+.photon-message-bar-action-button {
+  flex: none;
```
I made all buttons non-shrinkable

julienw commented on res/css/photon/message-bar.css:
```diff
@@ -5,53 +5,68 @@
+  flex: none;
   margin-left: auto;
   background: url(../../img/svg/close-dark.svg) no-repeat center / 16px 16px;
-  fill: currentColor;
```
This was a left-over from my previous experiments.

julienw commented on res/css/photon/message-bar.css:
```diff
@@ -5,53 +5,68 @@
-    var(--red-60);
+.photon-message-bar-error {
+  background-color: var(--red-60);
+  background-image: url(../../img/svg/error.svg);
```
This is simplified because most other properties are inherited. This makes it easier to change them in one place only.

julienw commented on res/css/photon/message-bar.css:
```diff
@@ -5,53 +5,68 @@
 }
 
-.photon-message-bar .photon-message-bar-close-button {
+.photon-message-bar-close-button {
```
I simplified this selector, it's more targeted now. It works fine because the specificity wasn't useful either.

julienw commented on res/css/photon/message-bar.css:
```diff
@@ -5,53 +5,68 @@
 
-.photon-message-bar .photon-button {
-  margin-left: 8px;
+.photon-message-bar-action-button {
```
It's more targeted and we don't need the specificity.

julienw commented on res/photon/index.html:
```diff
@@ -8,6 +8,7 @@
     <h1>Photon Styling Examples</h1>
     <p>This page has examples for the profiler’s implementation of the <a href="https://design.firefox.com/photon/">Photon design spec</a>.</p>
 
+    <h2>Photon Buttons</h2>
```
I added a few more titles to organize the page a bit more.

julienw commented on res/css/photon/message-bar.css:
```diff
@@ -71,6 +90,7 @@
 .photon-message-bar-warning .photon-message-bar-action-button:hover,
 .photon-message-bar-warning .photon-message-bar-close-button:hover:active {
   background-color: var(--yellow-70);
+  color: #fff;
```
I found that in the spec, the text color should be white because the background is darker.

julienw commented on res/photon/index.html:
```diff
@@ -132,76 +135,237 @@ <h3>Photon Radio Micro (Unofficial)</h3>
+          <button class="photon-button photon-message-bar-close-button" type="button"
+            aria-label="Hide the message" title="Hide the message">
+          </button>
+        </div>
```
This is the case I wanted to especially fix with this patch.

julienw commented:
> There are three sections of markup, but then six bars listed. Should we add all 6. 

Actually the markup is the same, the difference is the container. But maybe I should also show this example completely indeed...

> Also, it's a bit hard to read and know which is which. It might be nice to intersperse the markup and UI.

Yeah, I believe you're right.

> The margins around the small ones are different. Is this intentional?

Yeah, this is because with flex layout margins don't collapse anymore. I was a bit lazy to change that ;) Another way without flex is to use `width: max-content` directly on the elements, maybe I'll switch to this.

I'll handle all of this before landing!

julienw commented:
I removed the flex containers and replaced by using `width: max-content`. I believe it's better for this specific page. In the "real world" this won't always be useful to use this `width` rule so we shouldn't take this literarly, and that's why I didn't show it in the written code.

This also fixed the margin difference.

And finally I interspersed better the code and results.

I realized that in small conditions we're still not correct (the button should go below and aligned with the text, but the X should still stay at the top right). I wasn't sure how to fix this now and I don't need it now, so I'll just file a bug about that.
